mono/utils/mono-threads-coop.c

```diff
@@ -763,3 +763,26 @@ mono_threads_exit_no_safepoints_region (const char *func)
 	MONO_REQ_GC_UNSAFE_MODE;
 	mono_threads_transition_end_no_safepoints (mono_thread_info_current (), func);
 }
+
+void
+mono_thread_set_coop_aware (void)
+{
+	MONO_ENTER_GC_UNSAFE;
+	MonoThreadInfo *info = mono_thread_info_current_unchecked ();
+	if (info)
+		mono_atomic_store_i32 (&(info->coop_aware_thread), TRUE);
```

Comment:
I think we need a comment right here that it is critical that we do this store while in GC Unsafe mode.

Otherwise, if we allow the flag to be flipped from GC Safe, there could be a race if we're using signals for suspending: The suspend initiator could send a signal, but then the thread would switch to be coop aware right before it's preemptively suspended, and then I believe the resume code would use the cooperative resume instead of the signal-based preemptive resume.

Reply:
Added comment:

/* NOTE, this flag should only be changed while in unsafe mode. */
/* It will make sure we won't get an async preemptive suspend */
/* request against this thread while in the process of changing the flag */
/* affecting the threads suspend/resume behavior. */